signature/src/balrog.rs

```diff
@@ -0,0 +1,382 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/* This module handles verification of an X509 certificate chain and Mozilla's
+ * content signature algorithm which is used for signing and authentication of
+ * updates.
+ *
+ * This provides methods to parse an X509 certificate chain out of a PEM file,
+ * perform certificate chain validation to ensure the code signing certificate
+ * can be trusted, and verifies a content signature against the code signing
+ * certificate.
+ *
+ * The content signature itself is a base64-encoded ECDSA signature, using the
+ * content prefixed with "Content-Signature\x00" as the input data.
+ */
+
+use hex;
+use asn1_rs::ToDer;
+use ring::digest;
+use x509_parser::prelude::*;
+
+use oid_registry::{
+    OID_SIG_ECDSA_WITH_SHA256,
+    OID_SIG_ECDSA_WITH_SHA384,
+};
+
+pub struct Balrog<'a> { 
+    pub chain: Vec<X509Certificate<'a>>,
+    pub root_hash: Vec<u8>
+}
+
+// Errors that can be returned from the Balrog module.
+#[derive(Clone, Debug, PartialEq, thiserror::Error)]
+pub enum BalrogError {
+    #[error("Chain subject/issuer mismatch")]
+    ChainSubjectMismatch,
+    #[error("Certificate expired")]
+    CertificateExpired,
+    #[error("Issuer unauthorized for certificate signing")]
+    IssuerUnauthorized,
+    #[error("Certificate unauthorized for code signing")]
+    CodeSignUnauthorized,
+    #[error("Root hash parse failed")]
+    RootHashParseFailed,
+    #[error("Root hash mismatch")]
+    RootHashMismatch,
+    #[error("Hostname mismatch")]
+    HostnameMismatch,
+    #[error("Signature decoding failed")]
+    SignatureDecodeError,
+
+    #[error("PEM decoding error: {0:?}")]
+    PemDecodeError(String),
+    #[error("X509 validation failed: {0:?}")]
+    X509ValidationError(String),
+    #[error("X509 error: {0:?}")]
+    X509(X509Error)
+}
+
+impl From<X509Error> for BalrogError {
+    fn from(e: X509Error) -> Self {
+        BalrogError::X509(e)
+    }
+}
+
+impl From<PEMError> for BalrogError {
+    fn from(e: PEMError) -> Self {
+        BalrogError::PemDecodeError(e.to_string())
+    }
+}
+
+pub fn parse_pem_chain(input: &[u8]) -> Result<Vec<Pem>, PEMError> {
+    Pem::iter_from_buffer(input)
+        .collect::<Result<Vec<Pem>, PEMError>>()
+}
+
+/* A one-and-done function that parses a signature chain and validates a content signature. */
+pub fn parse_and_verify(
+    x5u: &[u8],
+    input: &[u8],
+    signature: &str,
+    current_time: i64,
+    root_hash: &str,
+    leaf_subject: &str
+) -> Result<(), BalrogError> {
+    let pem_chain = match parse_pem_chain(x5u) {
```

Comment:
I think all of these `match` statements can be replaced with `let pem_chain = match parse_pem_chain(x5u)?;` The `?` propagates the error.

Reply:
Ah cool! I didn't know about that language feature. That does clean things up a lot.